esinstall/src/entrypoints.ts

```diff
@@ -0,0 +1,268 @@
+    resolveExportMapEntry(entry?.default) ||
+    resolveExportMapEntry(entry?.require) ||
+    undefined
+  );
+}
+
+function resolveExportsMap(
+  packageManifest: PackageManifestWithExports,
+  exportsProp: string,
+): string | undefined {
+  const exportMapEntry = packageManifest.exports[exportsProp];
+  return resolveExportMapEntry(exportMapEntry);
+}
+
+type ResolveManifestFn = (dep: string, cwd: string) => [string | null, any];
+
+type ResolveEntrypointOptions = {
+  cwd: string;
+  packageLookupFields: string[];
+  resolveManifest?: ResolveManifestFn | undefined;
+};
+
+/**
+ * Resolve a "webDependencies" input value to the correct absolute file location.
+ * Supports both npm package names, and file paths relative to the node_modules directory.
+ * Follows logic similar to Node's resolution logic, but using a package.json's ESM "module"
+ * field instead of the CJS "main" field.
+ */
+export function resolveEntrypoint(
+  dep: string,
+  {cwd, packageLookupFields}: ResolveEntrypointOptions,
+): string {
+  // We first need to check for an export map in the package.json. If one exists, resolve to it.
+  const [packageName, packageEntrypoint] = parsePackageImportSpecifier(dep);
+  const [packageManifestLoc, packageManifest] = resolveDependencyManifest(packageName, cwd);
+
+  if (packageManifestLoc && packageManifest && typeof packageManifest.exports !== 'undefined') {
+    // If this is a non-main entry point
+    if (packageEntrypoint) {
+      const exportMapValue = resolveExportsMap(
+        packageManifest as PackageManifestWithExports,
+        './' + packageEntrypoint,
+      );
+
+      if (typeof exportMapValue !== 'string') {
+        throw new Error(
+          `Package "${packageName}" exists but package.json "exports" does not include entry for "./${packageEntrypoint}".`,
+        );
+      }
+      return path.join(packageManifestLoc, '..', exportMapValue);
+    } else {
+      const exportMapValue = resolveExportsMap(packageManifest as PackageManifestWithExports, '.');
+
+      if (exportMapValue) {
+        return path.join(packageManifestLoc, '..', exportMapValue);
+      }
+    }
+  }
+
+  // if, no export map and dep points directly to a file within a package, return that reference.
+  if (path.extname(dep) && !validatePackageName(dep).validForNewPackages) {
+    return fs.realpathSync.native(require.resolve(dep, {paths: [cwd]}));
+  }
+
+  // Otherwise, resolve directly to the dep specifier. Note that this supports both
+  // "package-name" & "package-name/some/path" where "package-name/some/path/package.json"
+  // exists at that lower path, that must be used to resolve. In that case, export
+  // maps should not be supported.
+  const [depManifestLoc, depManifest] = resolveDependencyManifest(dep, cwd);
+
+  if (!depManifest) {
+    try {
+      const maybeLoc = fs.realpathSync.native(require.resolve(dep, {paths: [cwd]}));
+      return maybeLoc;
+    } catch (err) {
+      // Oh well, was worth a try
```

Comment:
maybe logger.debug() this error?

Reply:
I think in this case it isn't actually an "error", this is just looking for a nested package.json just in case one exists. Maybe we should change the logic to not use try/catch.